File: hallucitron/openai_non_streaming.py
import json

import httpx

from hallucitron import openai_adapt
from hallucitron.hallu_structs import HalluStructuredRequest, HalluStructuredResult, HalluToolCall, HalluUsage, HalluApiError, dump_req_body


async def openai_non_streaming_call(req):
    http = httpx.AsyncClient()
    try:
        inp = await openai_adapt.adapt_messages(req.messages, http, req.prov_api_key, req.prov_endpoint, req.prov_name)
        assert req.prov_endpoint, "prov_endpoint must be set"
        endpoint = req.prov_endpoint.rstrip("/") + "/responses"
        body = {
            "model": req.provm_name,
            "input": inp,
        }
        if req.output_schema:
            body["text"] = {
                "format": {
                    "type": "json_schema",
                    "name": req.output_schema_name,
                    "schema": req.output_schema,
                    "strict": True,
                },
            }
        if req.tools:
            body["tools"] = req.tools
            if req.output_schema:
                body["tool_choice"] = "none"  # prevent tool calls (removing tools instead would break cache)
        if req.max_tokens > 0:
            body["max_output_tokens"] = req.max_tokens
        if req.reasoning_effort:
            effort = "high" if req.reasoning_effort == "max" else req.reasoning_effort
            body["reasoning"] = {"effort": effort}
        if req.temperature is not None and (not req.reasoning_effort or req.reasoning_effort == "none"):
            body["temperature"] = req.temperature
        dump_req_body(req, body)
        resp = await http.post(
            endpoint,
            headers={
                "Authorization": "Bearer %s" % req.prov_api_key,
                "content-type": "application/json",
            },
            json=body,
            timeout=120,
        )
        if resp.status_code >= 400:
            raise HalluApiError(resp.status_code, "openai-compat API error %d: %s" % (resp.status_code, resp.text))
        data = resp.json()
        actual_model = data.get("model", req.provm_name)
        response_id = data.get("id", "")
        raw_text = ""
        tool_calls = []
        for item in data.get("output", []):
            itype = item.get("type")
            if itype == "message":
                for part in item.get("content", []):
                    if part.get("type") == "output_text":
                        raw_text += part.get("text", "")
            elif itype == "function_call":
                name = item.get("name", "")
                call_id = item.get("call_id", "")
                args_str = item.get("arguments", "{}")
                arguments = json.loads(args_str)
                tool_calls.append(HalluToolCall(call_id=call_id, name=name, arguments=arguments))
        if not raw_text.strip() and not tool_calls:
            raise RuntimeError("openai_responses_api: no text output and no tool calls, max_tokens=%d, response=%s" % (req.max_tokens, json.dumps(data)))
        if tool_calls:
            parsed = None
        elif req.output_schema is None:
            parsed = raw_text
        else:
            parsed = json.loads(raw_text)
        usage = data.get("usage", {})
        provider_usage_json = usage
        input_tokens = usage.get("input_tokens", 0)
        cached_tokens = (usage.get("input_tokens_details") or {}).get("cached_tokens", 0)
        reasoning_tokens = (usage.get("output_tokens_details") or {}).get("reasoning_tokens", 0)
        cost_ticks = usage.get("cost_in_usd_ticks")
        provider_cost_usd = cost_ticks / 10_000_000_000.0 if cost_ticks is not None else None
        def tool_detail(name):
            return (usage.get("server_side_tool_usage_details") or {}).get(name, 0)
        return HalluStructuredResult(
            parsed=parsed,
            raw_text=raw_text,
            tool_calls=tool_calls,
            usage=HalluUsage(
                prompt_noncached=input_tokens - cached_tokens,
                output_tokens=usage.get("output_tokens", 0),
                including_reasoning_tokens=reasoning_tokens,
                cache_creation_input_tokens=0,
                cache_read_input_tokens=cached_tokens,
                call_web_search=tool_detail("web_search_calls"),
                call_x_search=tool_detail("x_search_calls"),
                call_code_interpreter=tool_detail("code_interpreter_calls"),
                call_document_search=tool_detail("document_search_calls"),
                call_file_search=tool_detail("file_search_calls"),
            ),
            provider_cost_usd=provider_cost_usd,
            actual_model=actual_model,
            stop_reason=data.get("status", ""),
            response_id=response_id,
            provider_usage_json=provider_usage_json,
        )
    finally:
        await http.aclose()

File: rust/src/openai_non_streaming.rs
use anyhow::{anyhow, Result};
use reqwest::Client;
use serde_json::{json, Value};

use super::openai_adapt;
use super::hallu_structs::{HalluStructuredRequest, HalluStructuredResult, HalluToolCall, HalluUsage};

// "thinking tokens streaming": you can’t stream raw thinking tokens. OpenAI’s docs say raw reasoning tokens are not visible via the API.
// The closest thing you can enable is a reasoning summary with reasoning.summary, and that summary comes back in a reasoning output item.
// struct OutputItem {
//   ...
//     summary: Option<Vec<ContentPart>>, // add this
// }

// include: ["reasoning.encrypted_content"] is for carrying reasoning across turns in stateless mode


pub async fn openai_structured_call(req: &HalluStructuredRequest) -> Result<HalluStructuredResult> {
    let http = Client::new();
    let input = openai_adapt::adapt_messages(&req.messages, &http, &req.prov_api_key, &req.prov_endpoint, &req.prov_name).await?;

    assert!(!req.prov_endpoint.is_empty(), "prov_endpoint must be set");
    let endpoint = format!("{}/responses", req.prov_endpoint.trim_end_matches('/'));

    let mut body = json!({
        "model": req.provm_name,
        "input": input,
    });
    if !req.output_schema.is_null() {
        body["text"] = json!({
            "format": {
                "type": "json_schema",
                "name": req.output_schema_name,
                "schema": req.output_schema,
                "strict": true,
            }
        });
    }
    if let Some(tools) = req.tools.as_array() {
        if !tools.is_empty() {
            body["tools"] = req.tools.clone();
            if !req.output_schema.is_null() {
                body["tool_choice"] = json!("none");  // prevent tool calls (removing tools instead would break cache)
            }
        }
    }
    if req.max_tokens > 0 {
        body["max_output_tokens"] = json!(req.max_tokens);
    }
    if !req.reasoning_effort.is_empty() {
        let effort = if req.reasoning_effort == "max" { "high" } else { req.reasoning_effort.as_str() };
        body["reasoning"] = json!({"effort": effort});
    }
    if let Some(t) = req.temperature {
        if req.reasoning_effort.is_empty() || req.reasoning_effort == "none" {
            body["temperature"] = json!(t);
        }
    }

    super::dump_req_body(req, &body);

    let response = http
        .post(&endpoint)
        .bearer_auth(&req.prov_api_key)
        .header("content-type", "application/json")
        .json(&body)
        .send()
        .await?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        anyhow::bail!("openai-compat API error {status}: {text}");
    }

    let resp: Value = response.json().await?;
    let actual_model = resp["model"].as_str().unwrap_or(&req.provm_name).to_string();
    let response_id = resp["id"].as_str().unwrap_or("").to_string();

    let mut raw_text = String::new();
    let mut tool_calls: Vec<HalluToolCall> = Vec::new();
    if let Some(output) = resp["output"].as_array() {
        for item in output {
            match item["type"].as_str() {
                Some("message") => {
                    if let Some(content) = item["content"].as_array() {
                        for part in content {
                            if part["type"].as_str() == Some("output_text") {
                                raw_text.push_str(part["text"].as_str().unwrap_or(""));
                            }
                        }
                    }
                }
                Some("function_call") => {
                    let name = item["name"].as_str().unwrap_or("").to_string();
                    let call_id = item["call_id"].as_str().unwrap_or("").to_string();
                    let args_str = item["arguments"].as_str().unwrap_or("{}");
                    let arguments: Value = serde_json::from_str(args_str).unwrap_or(json!({}));
                    tool_calls.push(HalluToolCall { call_id, name, arguments });
                }
                _ => {}
            }
        }
    }

    if raw_text.trim().is_empty() && tool_calls.is_empty() {
        anyhow::bail!("openai_responses_api: no text output and no tool calls, max_tokens={}, response={resp}", req.max_tokens);
    }

    let parsed: Value = if !tool_calls.is_empty() {
        Value::Null
    } else if req.output_schema.is_null() {
        Value::String(raw_text.clone())
    } else {
        serde_json::from_str(&raw_text)
            .map_err(|e| anyhow!("failed to parse structured JSON: {e}\nraw: {raw_text}"))?
    };

    let usage = &resp["usage"];
    let provider_usage_json = usage.clone();
    let input_tokens = usage["input_tokens"].as_u64().unwrap_or(0);
    let cached_tokens = usage.pointer("/input_tokens_details/cached_tokens")
        .and_then(Value::as_u64).unwrap_or(0);
    let reasoning_tokens = usage.pointer("/output_tokens_details/reasoning_tokens")
        .and_then(Value::as_u64).unwrap_or(0);
    let provider_cost_usd = usage["cost_in_usd_ticks"].as_f64().map(|t| t / 10_000_000_000.0);
    let tool_detail = |name: &str| -> u64 {
        usage.pointer(&format!("/server_side_tool_usage_details/{name}"))
            .and_then(Value::as_u64).unwrap_or(0)
    };
    Ok(HalluStructuredResult {
        parsed,
        raw_text,
        thinking_text: String::new(),
        provider_specific_stuff: Value::Null,
        tool_calls,
        usage: HalluUsage {
            prompt_noncached: input_tokens.saturating_sub(cached_tokens),
            output_tokens: usage["output_tokens"].as_u64().unwrap_or(0),
            including_reasoning_tokens: reasoning_tokens,
            cache_creation_input_tokens: 0,
            cache_read_input_tokens: cached_tokens,
            call_web_search: tool_detail("web_search_calls"),
            call_x_search: tool_detail("x_search_calls"),
            call_code_interpreter: tool_detail("code_interpreter_calls"),
            call_document_search: tool_detail("document_search_calls"),
            call_file_search: tool_detail("file_search_calls"),
            ..Default::default()
        },
        coins: 0,
        price_breakdown: Vec::new(),
        provider_cost_usd,
        actual_model,
        stop_reason: resp["status"].as_str().unwrap_or("").to_string(),
        response_id,
        provider_usage_json,
        sse_log: Vec::new(),
    })
}
